### PR TITLE
Add categories

### PR DESCRIPTION
**Pull request by SashaShostyr**

### CODE REVIEW

File: app/views/items/index.html.slim [hunk 1]
```diff
@@ -5,13 +5,15 @@ table.table
...
       th = 'Description'
+      th = 'Category'
       th = sort_link 'created_at', 'Created On'
       th colspan="3"
   tbody
     - @items.each do |item|
       tr
         td= item.title
         td= item.description
+        td= item.category&.name
```
Comment (Sergyenko) [inline app/views/items/index.html.slim:16]: item.category&.name - typo?

**Review by Sergyenko — CHANGES_REQUESTED** [review]

File: app/controllers/categories_controller.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    end
+  end
+
+  def destroy
+    @category.destroy
+    redirect_to categories_url
+  end
+
+  private
+
```
Comment (Sergyenko) [inline app/controllers/categories_controller.rb]: Think about the creating helper for this code. As it's the same as here:
https://github.com/cybergizer/nicher/pull/26/files#diff-dbc6fdbfa3706e8655ca693ae7639d6bR54

**Review by Sergyenko — DISMISSED** [review]

File: app/views/items/index.html.slim (hunk 1 above)
Comment (SashaShostyr) [reply app/views/items/index.html.slim:16]: no, & - safe navigation operator

File: app/controllers/items_controller.rb [hunk 3]
```diff
@@ -56,6 +56,6 @@ def set_item
   end
 
   def item_params
-    params.require(:item).permit(:title, :description)
+    params.require(:item).permit(:title, :description, :category_id)
```
Comment (anatoliliotych) [inline app/controllers/items_controller.rb]: be sure that you will follow the same approach @Vadser is doing

File: spec/controllers/categories_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+RSpec.describe CategoriesController, type: :controller do
+  login_user
+
+  let(:valid_attributes) {
+    {
+      name: 'Clothes',
+    }
+  }
+
+  let(:invalid_attributes) {
```
Comment (anatoliliotych) [inline spec/controllers/categories_controller_spec.rb]: why it's not valid?

File: spec/controllers/categories_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    {
+      parent_id: nil
+    }
+  }
+
+  let(:valid_session) { {} }
+
+  describe 'GET #index' do
+    it 'returns a success response' do
+      Category.create! valid_attributes
```
Comment (anatoliliotych) [inline spec/controllers/categories_controller_spec.rb]: please, talk to @Vadser about creation of objects.

**Review by anatoliliotych — CHANGES_REQUESTED** [review]

File: app/helpers/categories_helper.rb [hunk 6]
```diff
@@ -0,0 +1,2 @@
+module CategoriesHelper
```
Comment (anatoliliotych) [inline app/helpers/categories_helper.rb]: remove helper it's empty

File: app/helpers/application_helper.rb [hunk 7]
```diff
@@ -1,2 +1,15 @@
 module ApplicationHelper
+  def set_resource
```
Comment (anatoliliotych) [inline app/helpers/application_helper.rb]: move all methods to concern in a separate PR.

File: app/views/items/_form.html.slim [hunk 8]
```diff
@@ -13,5 +13,8 @@
   .form-group
     = form.label :description
     = form.text_area :description
+  .form-group
+    = form.label :category_id
+    = form.collection_select(:category_id, Category.all, :id, :name, :include_blank => "Please select")
```
Comment (anatoliliotych) [inline app/views/items/_form.html.slim]: `Category.all` - will show you categories from other users!

File: spec/controllers/categories_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,146 @@
+require 'rails_helper'
+
+RSpec.describe CategoriesController, type: :controller do
+  login_user
+
+  let(:category) { create(:category) }
+  let(:valid_attributes) { attributes_for(:category) }
+  let(:invalid_attributes) { attributes_for(:invalid_category) }
+
+  describe 'GET #index' do
```
Comment (anatoliliotych) [inline spec/controllers/categories_controller_spec.rb]: add test for correct route

Comment (anatoliliotych) [reply spec/controllers/categories_controller_spec.rb]: add this for all actions!

File: spec/controllers/categories_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+    context 'with valid params' do
+      it 'creates a new Category' do
+        expect {
+          post :create, params: { category: valid_attributes }
+        }.to change(Category, :count).by(1)
+      end
+
+      it 'redirects to the created category' do
+        post :create, params: { category: valid_attributes }
+        expect(response).to redirect_to(Category.last)
```
Comment (anatoliliotych) [inline spec/controllers/categories_controller_spec.rb]: bad approach! check that you are redirected exactly to created category

File: spec/controllers/categories_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+        put :update, params: { id: category, category: invalid_attributes }
+      end
+
+      it "does not change category's attributes" do
+        category.reload
+        category.name.should_not be_nil
+      end
+
+      it "re-renders the edit method " do
+        expect(response).to render_template :edit
```
Comment (anatoliliotych) [inline spec/controllers/categories_controller_spec.rb]: check that flash message is shown for all cases!

File: spec/models/item_spec.rb [hunk 12]
```diff
@@ -1,5 +1,24 @@
 require 'rails_helper'
 
 RSpec.describe Item, type: :model do
-  pending "add some examples to (or delete) #{__FILE__}"
+  let(:user) { User.create(email: 'test@gmail.com', password: 'password', password_confirmation: 'password') }
```
Comment (anatoliliotych) [inline spec/models/item_spec.rb]: use FactoryBot!

File: spec/spec_helper.rb [hunk 13]
```diff
@@ -26,6 +26,7 @@
     # ...rather than:
     #     # => "be bigger than 2"
     expectations.include_chain_clauses_in_custom_matcher_descriptions = true
+    expectations.syntax = [:should, :expect]
```
Comment (anatoliliotych) [inline spec/spec_helper.rb]: use only expects!

**Review by anatoliliotych — CHANGES_REQUESTED** [review]

File: app/controllers/categories_controller.rb [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+  before_action :set_category, only: %i[show edit update destroy]
+
+  def index
+    @categories = current_user.categories.roots
+  end
+
+  def show; end
+
+  def new
+    @category = current_user.categories.create
```
Comment (anatoliliotych) [inline app/controllers/categories_controller.rb]: you shouldn't create db record on page opening!

File: spec/controllers/categories_controller_spec.rb [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+      it "does not save the new category" do
+        expect{
+          post :create, params: { category: invalid_attributes }
+        }.to_not change(Category,:count)
+      end
+
+      it "re-renders the new method" do
+        post :create, params: { category: invalid_attributes }
+        category = assigns(:category)
+        expect(category.errors).to be_present
```
Comment (anatoliliotych) [inline spec/controllers/categories_controller_spec.rb]: not exact check, you can get validation error not related to what you've meant, so be more specific in tests.

**Review by anatoliliotych — CHANGES_REQUESTED** [review]

File: app/controllers/categories_controller.rb [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+  def new
+    @category = current_user.categories.build
+  end
+
+  def edit; end
+
+  def create
+    @category = current_user.categories.create(category_params)
+
+    if @category.save
```
Comment (anatoliliotych) [inline app/controllers/categories_controller.rb]: replace with valid? or remove create and use new

File: app/models/item.rb [hunk 17]
```diff
@@ -1,8 +1,11 @@
 class Item < ApplicationRecord
-  PARAMS = %i[title description niche_id].freeze
+  PARAMS = %i[title description niche_id category_id].freeze
 
   belongs_to :user
+
```
Comment (anatoliliotych) [inline app/models/item.rb]: use the same sections together

**Review by anatoliliotych — CHANGES_REQUESTED** [review]

**Review by anatoliliotych — APPROVED** [review]

**Comment by anatoliliotych** [comment]
lgtm

**Pull request merged**